File: N.Maximum_Subarray.js
//53. Maximum Subarray
var maxSubArray = function (nums) {
  //[5,3,-11,9]
  if (nums == null || nums.length == 0) return -Infinity;
  let max = nums[0]; //5

  //initiate dp this dp store single val not array since only need maximum val
  let dp = nums[0]; //5

  for (let i = 1; i < nums.length; i++) {
    dp = Math.max(dp + nums[i], nums[i]); //5+3 > 3; 8-11 > -11; -3<9
    //update dp and max in the meantime
    max = Math.max(dp, max);
  }
  return max;
};

File: dp/N.Maximum_Subarray.py
'''
53. Maximum Subarray
    LOGIC:
    1.dp - single val DP
    2.need both dp val and max_result! WHY? dp is for compare add cur, or not, if +cur> cur, will extend subarray;otherwise, start over subarray
'''
def maxSubArray(self, nums: List[int]) -> int:
        max_v = nums[0] #has negative
        dp = nums[0]
        
        for i in range(1,len(nums)):
            dp = max(dp+nums[i], nums[i])
            max_v = max(dp, max_v)
        return max_v

'''
918. Maximum Sum Circular Subarray
'''